src/util/atomic_io.c

```diff
@@ -58,3 +60,35 @@ ssize_t sss_atomic_io_s(int fd, void *buf, size_t n, bool do_read)
 
     return pos;
 }
+
+ssize_t sss_atomic_write_safe_s(int fd, void *buf, size_t len)
+{
+    uint32_t ulen = len;
+    ssize_t ret;
+
+    ret = sss_atomic_write_s(fd, &ulen, sizeof(uint32_t));
+    if (ret != sizeof(uint32_t)) {
+        errno = errno == 0 ? EIO : errno;
+        return -1;
+    }
+
+    return sss_atomic_write_s(fd, buf, len);
+}
+
+ssize_t sss_atomic_read_safe_s(int fd, void *buf, size_t buf_len, size_t *_len)
+{
+    uint32_t ulen;
+    ssize_t ret;
+
+    ret = sss_atomic_read_s(fd, &ulen, sizeof(uint32_t));
+    if (ret != sizeof(uint32_t)) {
+        errno = errno == 0 ? EIO : errno;
+        return -1;
+    }
+
```

Comment:
Hi,

Coverity complains:

```
[Error: TAINTED_SCALAR (]()[CWE-20](https://cwe.mitre.org/data/definitions/20.html)): [[#def6]](https://cov01.lab.eng.brq.redhat.com/covscanhub/task/242632/log/added.html#def6)
sssd-2.6.2/src/util/atomic_io.c:83: tainted_argument: Calling function "sss_atomic_io_s" taints argument "ulen".
sssd-2.6.2/src/util/atomic_io.c:93: tainted_data: Passing tainted expression "ulen" to "sss_atomic_io_s", which uses it as a loop boundary.
sssd-2.6.2/src/util/atomic_io.c:93: remediation: Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.
#   91|       }
#   92|   
#   93|->     return sss_atomic_read_s(fd, buf, ulen);
#   94|   }

[Error: TAINTED_SCALAR (]()[CWE-20](https://cwe.mitre.org/data/definitions/20.html)): [[#def7]](https://cov01.lab.eng.brq.redhat.com/covscanhub/task/242632/log/added.html#def7)
sssd-2.6.2/src/util/child_common.c:524: tainted_return_value: Function "sss_atomic_read_safe_s" returns tainted data.
sssd-2.6.2/src/util/child_common.c:524: var_assign: Assigning: "size" = "sss_atomic_read_safe_s(state->fd, buf, 256UL, &len)", which taints "size".
sssd-2.6.2/src/util/child_common.c:535: lower_bounds: Checking lower bounds of signed scalar "size" by taking the true branch of "size > 0L".
sssd-2.6.2/src/util/child_common.c:543: tainted_data: Passing tainted expression "size" to "safealign_memcpy", which uses it as an offset.
sssd-2.6.2/src/util/child_common.c:543: remediation: Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.
#  541|           }
#  542|   
#  543|->         safealign_memcpy(&state->buf[state->len], buf,
#  544|                            size, &state->len);
#  545|
```

So maybe it would make sense to add a `max` parameter to `sss_atomic_read_safe_s()` where the caller can specify its expectations about the data to be read. As an alternative a hard limit of e.g. 1MB should work as well.

bye,
Sumit

Reply:
added if (ulen > buf_len) return ERANGE